Update to 4.2.0

porras opened the pull request:
Hi! I'm not exactly a flatpak expert (first time twiddling around) but this _Works In My Machine (tm)_ (Fedora 37).

I also fixed the indentation (which was somehow mixing spaces and tabs) but I did it in a separate commit to make the review easier (although using the UI option to hide whitespace changes also works I guess). I'm adding comments about each of the changes, in the diff.

hfiguiere committed:
WiP: Update to SonicPi 4

- Added Elixir and some dependencies
- Added C++ deps

porras commented on net.sonic_pi.SonicPi.json:
```diff
@@ -40,45 +40,56 @@
+                "--without-reltool",
+                "--without-ssh",
+                "--without-syntax_tools",
+                "--without-tftp"
```
`tools` and `parsetools` are needed to compile Elixir. `sasl` is required to compile the Sonic Pi Elixir app (tau).

porras commented on net.sonic_pi.SonicPi.json:
```diff
@@ -40,45 +40,56 @@
+                    "sha256": "6cece7317044233996e6056e86a09e8ba25971c5f60130738d0b6baccb3b99e0"
                 }
             ]
         },
```
New Sonic Pi requirement.

porras commented on net.sonic_pi.SonicPi.json:
```diff
@@ -154,10 +165,13 @@
-                "cd app && ./linux-prebuild.sh --build-aubio",
-                "cd app && ./linux-config.sh --config RelWithDebInfo",
-                "cmake --build app/build"
+                "cd app && ./linux-build-all.sh --config RelWithDebInfo && ./linux-release.sh"
```
From what I know, the whole build process has been revamped and this simplifies this section, it simply works with the provided scripts (I left the `--config RelWithDebInfo` option around; I don't know exactly what it does, but I checked that it gets passed through).

porras commented on net.sonic_pi.SonicPi.json:
```diff
@@ -154,10 +165,13 @@
+                "build-args": [
+                    "--share=network"
+                ]
+            },
```
The new build process downloads the dependencies (that's what makes it simpler in part, I guess) using `vcpkg`, so the network needs to be available during build.

hfiguiere replied:
THIS IS NOT ALLOWED.

I have tried for weeks to get something working. But offline doesn't work.

porras replied:
I see now that you can't do this in the sandbox environment (the build failed because of that https://buildbot.flathub.org/#/builders/19/builds/13456)... I'll try to get around it, but any idea is welcome!

hfiguiere replied:
this #12. I hadn't pushed it because it didn't work, so I just did.

porras commented on net.sonic_pi.SonicPi.json:
```diff
@@ -185,17 +199,7 @@
-                "cp -R app/server/ruby/vendor $FLATPAK_DEST/app/server/ruby",
-                "cp -R app/server/native $FLATPAK_DEST/app/server",
-                "install -Dm644 -t $FLATPAK_DEST/app/server/erlang/sonic_pi_server/ebin app/server/erlang/sonic_pi_server/ebin/*.beam",
+                "cd app/build/linux_dist && cp -av * $FLATPAK_DEST",
```
There is a new script that prepares a new release with all the needed files in the right place, so instead of manually going one by one like before (some of the stuff has changed so it didn't work _as is_), I thought it's just better run that script (in the build section) and here simply copy the whole result into the tree.

The only ones I left unchanged are the flatpak specific files (icons, desktop file, etc.).

porras commented on net.sonic_pi.SonicPi.json:
```diff
@@ -205,16 +209,12 @@
-                    "type": "patch",
-                    "path": "sonic-pi-fixes.patch"
+                    "url": "https://github.com/sonic-pi-net/sonic-pi/archive/refs/tags/v4.2.0.tar.gz",
+                    "sha256": "22376b5acf528582c6b8f43a6c53df368775ac4cc7b1136fdfc073c05d2ad56d"
```
This is what it all is about! :smile_cat:

porras commented on net.sonic_pi.SonicPi.json:
```diff
@@ -205,16 +209,12 @@
                 {
                     "type": "patch",
-                    "path": "sonic-pi-sc-path.patch"
+                    "path": "sonic-pi-no-libstdc++.patch"
```
From what I could gather, the patches that were previously here are no longer needed.

I needed to add a new one though (see comment below).

porras commented on sonic-pi-no-libstdc++.patch:
```diff
@@ -0,0 +1,12 @@
+-        stdc++fs
+         rt
+     PRIVATE
+         crossguid
```
This is the thing that had me puzzled for a while because I know nothing about C++ :sweat_smile: 

From what I could understand via some googling, etc., is that with gcc >= 9 (which is a bit old and the flatpak in any case doesn't use), linking this library is not necessary. In general, it is also harmless because it's present in most systems, but it turns out, in the flatpak sdk it is not. I'll also submit that patch to the Sonic Pi project, but it is a bit more involved than here because there gcc 9 needs to be supported I guess so I need to learn how to write that condition in CMake (I've see it's possible).

hfiguiere replied:
it exist because there are still too many people runing antiquate distro like Ubuntu LTS. Flatpak doesn't ship the .a stub while distro do, so it fails. It's a recurring problem in a world of developer that think Linux == Ubuntu

porras reviewed (commented):
I hope this comments help!

hfiguiere commented:
See #12 if you can fix it.

porras commented:
@hfiguiere I see you got further than me, but still hit the wall :sob:

hfiguiere commented:
using obscures platforms lead to that kind of brick wall.

porras commented:
I just pushed something that seems to work in my machine (and without `--share=network`). I followed an approach slightly different to #12. Instead of bypassing the `vcpkg` and `hex` tools (after seeing it would be a lot of work) I decided to keep using them, but previously download the files and place them in the cache directories where those tools expect to find them. Of course creating the list of files to download is quite tedious, so I created two simple scripts that do it. They are not needed at build time, but I added them in the PR as documentation (and to reuse them for future releases). After a bit of back and forth, it got to successfully build :)

There are a couple of questionable things that are probably not good flatpak practice:

* It downloads two binary distributions from their repos (`vcpkg-glibc` and `rebar3`)
* There is a binary file added in the repo (it's not really binary, it's some kind of erlang specific serialization format) generated in my local machine (containing the list of cached dependencies)

The first one can be probably worked around and get to build those from source. The second I wasn't able to get rid of (although I know almost nothing about erlang/elixir tools, and it might be possible). `mix` wouldn't _see_ the cache files if this file is not there :(

There are still some things I want to cleanup (and also upgrade erlang, ruby, and the rest of sonic pi dependencies, which is not required but nice I guess) but I wanted to push this to 1) get the bot to test it and 2) get your feedback @hfiguiere. I'd appreciate it!

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -150,14 +161,42 @@
             ]
         },
+        {
+            "name": "cmake",
```
why do you need cmake? it's in the runtime.

porras replied:
it's sure overkill to have to compile this, but `vcpkg` requires this version (the one in the runtime is the previous one).

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -205,16 +234,345 @@
+                },
+                {
+                    "type": "file",
+                    "url": "https://gitlab.xiph.org/xiph/ogg/-/archive/v1.3.5/ogg-v1.3.5.tar.gz",
```
this is in the runtime.

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -205,16 +234,345 @@
+                },
+                {
+                    "type": "file",
+                    "url": "https://github.com/xiph/flac/archive/1151c93e992bb8c7c6394e04aa880d711c531c7f.tar.gz",
```
flac is in the runtime

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -205,16 +234,345 @@
+                },
+                {
+                    "type": "file",
+                    "url": "https://github.com/xiph/vorbis/archive/v1.3.7.tar.gz",
```
vorbis is in the runtime

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -205,16 +234,345 @@
+                    "type": "file",
+                    "url": "https://github.com/xiph/opus/archive/5c94ec3205c30171ffd01056f5b4622b7c0ab54c.tar.gz",
+                    "dest": "app/vcpkg/downloads",
+                    "dest-filename": "xiph-opus-5c94ec3205c30171ffd01056f5b4622b7c0ab54c.tar.gz",
```
opus is in the runtime.

hfiguiere commented on scripts/vcpkg-deps.rb:
```diff
@@ -0,0 +1,30 @@
+# Script to generate the list of downloads for vcpkg dependencies.
```
could this be written in a language we all understand? flathub uses python in general.

porras replied:
sure, I never wrote python (more than hello world anyway) but translating such simple scripts should be easy.

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -150,14 +161,42 @@
-                "cd app && ./linux-config.sh --config RelWithDebInfo",
-                "cmake --build app/build"
+                /* make downloaded vcpkg binary executable */
+                "chmod +x app/vcpkg/vcpkg",
```
except we want to build on aarch64, so this is NO NO.

porras replied:
I honestly didn't know that other architectures than x86 are supported, obviously this can't be done this way. It can be built from source and I was planning to do that change, I just took the shortcut and left it for later to deal first with the elixir part, which is the one I wasn't sure was possible at all. I'll fix this.

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -205,16 +234,345 @@
+                },
+                {
+                    "type": "file",
+                    "url": "https://github.com/microsoft/vcpkg-tool/releases/download/2022-07-21/vcpkg-glibc",
```
it's lovely to see that there is no mention of even the architecture the binary is meant to run on.

Note: we still build on aarch64.

porras commented:
@hfiguiere thanks for the quick feedback! I'll see how I can deal with those issues.

Sonic Pi is a super cool project, but it's proving to be... challenging to package, with its mix of platforms. I have zero experience with this kind of environments and platforms (personally I have only worked with dynamic languages, which have their own challenges, but they're quite different).

porras committed:
Update to v4.2.0

Includes:

* Update Sonic Pi to 4.2.0
* Update Elixir to 1.14
* Update Ruby to 3.1.2
* Remove ruby-ffi (no longer needed)
* Use the provided linux-release.sh script to generate the directory structure
with all the needed files and then just copy the whole thing (instead of
individually listed paths)
* Different approach to Elixir/Erlang dependencies (this WIP because there are a
couple of not-nice things but already works). These are the current
changes (with which I managed to successfuly build) that I intend to try to
refactor to make them nicer:
  * Move everything to the sonicpi module build
  * Use compiled releases of rebar3 and hex (these are multiplatform because
  they're not binary, they're Erlang VM files, but still not nice)
  * Place downloaded tarballs where hex expects to find them (a cache directory)
  * Sadly this means that the cache.ets file (generated locally) is needed as a
  "manifest" of those cached files. This is an Erlang specific serialization format

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -205,16 +336,253 @@
+                    "dest-filename": "rebar3",
+                    "sha256": "9c9d0fa142e60711147d23de7520fab4a8e32608fad5fddfbcb764e4ce98505f"
+                },
+                /* elixir dependencies */
```
if these are generated by a script then it should be in a separate .json. Easier for maintenance.

hfiguiere replied:
also were are the scripts?

porras commented:
@hfiguiere it's still work in progress but thanks for the quick look. I rebased on your #12 since you had already dealt with the C++ stuff so it was a bit silly that I would also do it.

I still need to translates the scripts into python, and improve some other couple of things.

porras committed:
Move elixir dependencies to a separate file

And a python script to generate it

hfiguiere commented on scripts/elixir-deps.py:
```diff
@@ -0,0 +1,52 @@
+            'dest': ".hex/packages/hexpm",
+            'sha256': hashlib.sha256(Path(filename).read_bytes()).hexdigest()
+    } for filename in glob.glob('packages/**/*.tar')
+]
```
shouldn't this also add the `cache.ets` entry since it's part of the artifacts? (instead of having it in the main manifest)

porras replied:
Hi @hfiguiere, I don't know if you saw that I already pushed that change you propose, some days ago. Let me know if you find the solution ok. Thanks!

porras commented:
>
> shouldn't this also add the cache.ets entry since it's part of the
> artifacts? (instead of having it in the main manifest)
>
That makes sense, I'll do that as soon as I'm back on a computer (it'll be
in some days).

Other than that, is this fit to merge?

>

hfiguiere commented:
bot, build

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -137,26 +143,161 @@
             "buildsystem": "simple",
             "build-commands": [
-                "gem install --local ffi-1.13.1.gem"
+                "./waf configure --prefix=$FLATPAK_DEST build",
```
```suggestion
                "./waf configure --prefix=$FLATPAK_DEST build -j $FLATPAK_BUILDER_N_JOBS",
```

so that we can parallel build

hfiguiere commented:
I'll merge this. Thanks!

Close #12

porras commented:
> I'll merge this. Thanks!

No, thanks to you!

The build of the merge seems to have failed in one of the architectures due to what I'd say it's a network issue (fetching the screenshot). Maybe someone can retrigger it, or will it be retried? https://buildbot.flathub.org/#/apps/net.sonic_pi.SonicPi

hfiguiere commented:
I'll retrigger.

I also have a 22.08 runtime update in the pipe, and 4.3 release.

hfiguiere commented:
btw do you want to be comaintainer of this flatpak?

porras commented:
Sure, I'll be happy to do that. I guess it involves keeping track of upstream releases and updating the flatpak, taking care of possible issues. What this PR does (and which _should_ be easier in the coming releases).

The build failed again, this time both :disappointed:

```
flatpak run --env=G_DEBUG=fatal-criticals --command=appstream-util org.flatpak.Builder validate builddir/*/share/appdata/net.sonic_pi.SonicPi.appdata.xml
 in dir /srv/buildbot/worker/build-x86_64-7/build (timeout 1200 secs)
 watching logfiles {}
 argv: b'flatpak run --env=G_DEBUG=fatal-criticals --command=appstream-util org.flatpak.Builder validate builddir/*/share/appdata/net.sonic_pi.SonicPi.appdata.xml'
 using PTY: False
builddir/files/share/appdata/net.sonic_pi.SonicPi.appdata.xml: FAILED:
• url-not-found         : <screenshot> url not valid [https://sonic-pi.net/media/images/tutorial/GUI.png]: Resolving timed out after 5000 milliseconds
Validation of files failed
program finished with exit code 1
elapsedTime=22.703965
```

It's just trying to fetch this [image](https://sonic-pi.net/media/images/tutorial/GUI.png), which from my side loads without problem. I have no idea what the problem might be.

hfiguiere commented:
@barthalion can you invite Sergio ( @porras  ) ? Thanks.

barthalion commented:
I've just sent an invite to @porras.

porras commented:
Thanks!
